controllers/osbuildconfig_controller.go

```diff
@@ -125,11 +147,134 @@ func (r *OSBuildConfigReconciler) createNewOSBuildCR(ctx context.Context, osBuil
+			return err
+		}
+	}
+
 	logger.Info("A new OSBuild CR was created", "OSBuild", osBuild.Name)
 
 	return nil
 }
 
+func (r *OSBuildConfigReconciler) applyTemplate(ctx context.Context, osBuildConfig *osbuilderprojectflottaiov1alpha1.OSBuildConfig, osBuildConfigSpecDetails *osbuilderprojectflottaiov1alpha1.BuildDetails, osBuildName string, osBuild *osbuilderprojectflottaiov1alpha1.OSBuild) (*v1.ConfigMap, *osbuilderprojectflottaiov1alpha1.OSBuildConfigTemplate, error) {
+	var kickstartConfigMap *v1.ConfigMap
+	var osConfigTemplate *osbuilderprojectflottaiov1alpha1.OSBuildConfigTemplate
+	buildConfigCustomizations := osBuildConfigSpecDetails.Customizations
+	if template := osBuildConfig.Spec.Template; template != nil {
+		var err error
+		osConfigTemplate, err = r.OSBuildConfigTemplateRepository.Read(ctx, template.OSBuildConfigTemplateRef, osBuildConfig.Namespace)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		mergeCustomization(osConfigTemplate, osBuildConfigSpecDetails, buildConfigCustomizations)
+
+		kickstartConfigMap, err = r.createKickstartConfigMap(ctx, osBuildConfig, osConfigTemplate, osBuildName, osBuild.Namespace)
+		if err != nil {
+			return nil, nil, err
+		}
+		if kickstartConfigMap != nil {
+			osBuild.Spec.Kickstart = &osbuilderprojectflottaiov1alpha1.NameRef{Name: osBuildName}
+		}
+	}
+	return kickstartConfigMap, osConfigTemplate, nil
+}
+
+func (r *OSBuildConfigReconciler) setKickstartConfigMapOwner(ctx context.Context, kickstartConfigMap *v1.ConfigMap, osBuild *osbuilderprojectflottaiov1alpha1.OSBuild) error {
+	patch := client.MergeFrom(kickstartConfigMap)
+	err := controllerutil.SetOwnerReference(osBuild, kickstartConfigMap, r.Scheme)
+	if err != nil {
+		return err
+	}
+	return r.Client.Patch(ctx, kickstartConfigMap, patch)
+}
+
+func (r *OSBuildConfigReconciler) createKickstartConfigMap(ctx context.Context, osBuildConfig *osbuilderprojectflottaiov1alpha1.OSBuildConfig, osConfigTemplate *osbuilderprojectflottaiov1alpha1.OSBuildConfigTemplate, name, namespace string) (*v1.ConfigMap, error) {
+	cm := &v1.ConfigMap{}
+	err := r.Client.Get(ctx, types.NamespacedName{Namespace: namespace, Name: name}, cm)
+	if err == nil {
+		// CM has already been created, returning it
+		return cm, nil
+	}
+	if !errors.IsNotFound(err) {
+		return nil, err
+	}
+
+	kickstart, err := r.getKickstart(ctx, osConfigTemplate, osBuildConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if kickstart != nil {
+		cm = &v1.ConfigMap{
+			ObjectMeta: ctrl.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+			Data: map[string]string{
+				"kickstart": *kickstart,
+			},
+		}
+
+		err = r.Client.Create(ctx, cm)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return cm, nil
+}
+
+func mergeCustomization(osConfigTemplate *osbuilderprojectflottaiov1alpha1.OSBuildConfigTemplate, osBuildConfigSpecDetails *osbuilderprojectflottaiov1alpha1.BuildDetails, buildConfigCustomizations *osbuilderprojectflottaiov1alpha1.Customizations) {
```

Comment:
Not sure I understand what this method will produce. Maybe we weren't clear about the expected output either.
For example, we would want to allow the user to add packages at the `OSBuildConfig` level to the ones defines at the `OSBuildConfigTemplate` level. E.g. if the `OSBuildConfigTemplate` sets packages `a` and `b`, the `OSBuildConfig` should be able to add package `c`. So, there should be a smart merge here (i.e. merge and unique).
- In `Packages` it is simple as it is a list of strings.
-  In Users we need to check the user's name and if exists on both, the `OSBuildConfig` takes precedence. 
- Services is the most complicated one as we need to make sure the `enabled` and `disabled` lists do not contradict each other. But the rule stays the same the `OSBuildConfig` takes precedence

Reply:
I understood from our previous chat, that `OSBuildConfig` takes precedence and so it does here, but at the level of whole sets of customizations - packages, services and users. For example - if packages are configured in both template and build config, only packages from build config would be used. If that's not the desired behavior, I will rework it.

Reply:
Done.